### DISCUSSION TITLE
Merging Object & Record

### DISCUSSION THREAD

**Discussion by jonathanstanley** [colinhacks/zod#3723] [Q&A]
I'd like to parse a spreadsheet provided by users. A product spreadsheet could look like this:

| Title | Price | Published | Attribute1 | Attribute2 |
| :------------- | :------------- | :------------- | :------------- | :------------- |
| Balloons  | 9.99 | false | latex | neon |
| Cake | 20.00 | true | flour and sugar | 1 lb |

I want to allow users to upload any number of attributes (ex: Attribute3, Attribute4, ...Attribute999).

individually, the schemas are simple enough

```ts
let staticColSchema = z.object({
  Title:z.string(),
  Price:z.number(),
  Published:z.boolean(),
});

let dynamicColSchema = z.record(z.string().startsWith("Attribute"),z.string());
```

however, I need a schema for the entire spreadsheet

```
let spreadsheetSchema = staticColSchema.and(dynamicColSchema).array()
```

Doesn't work. For example, 

* `record` will not strip out invalid keys; instead if fails.
* `catchall` cannot specify keys
* `intersection` doesn't work because not all columns have that same record type
* 
Possible solutions I've thought about:

1. allow records to strip out invalid keys instead of failing?
2. https://github.com/colinhacks/zod/issues/2195
3. https://github.com/colinhacks/zod/issues/1333
4. https://stackoverflow.com/questions/75546547/create-a-schema-that-has-some-dynamic-keys-and-some-static-keys-with-zod
5. maybe same issue: https://github.com/colinhacks/zod/issues/2200#issuecomment-2005242701
6. https://stackoverflow.com/questions/73459017/custom-validation-of-optional-keys-in-zod

**Comment by jonathanstanley** [comment]
The best I can think of:

```ts
let staticColSchema = z.object({
	Title: z.string(),
	Price: z.number(),
	Published: z.boolean()
});

let uploadSchema = staticColSchema
	.passthrough()
	.transform((row) => {
		return Object.fromEntries(
			// use filter to strip disallowed properties
			Object.entries(row).filter(
				([key, value]) =>
					// allow all statically defined elements
					Object.keys(staticColSchema.shape).includes(key) ||
					// allow properties that have valid title and value type
					(key.startsWith('Attribute') && z.string().safeParse(value).success)
			)
		);
	})
	.array();
```

... but the parsed type is bad

**Reply by sunnylost** [reply]
You can simplify your code to make it better:

```ts
const spreadsheetSchema = staticColSchema
  .passthrough()
  .transform(({ Title, Price, Published, ...rest }) => ({
    ...rest,
  }))
  .pipe(dynamicColSchema)
  .array();
```
